File: packages/typescript/src/features/advanced-capabilities/actions/unmuteRoom.ts
import { requireActionSpec } from "../../../generated/spec-helpers.ts";
import {
	findKeywordTermMatch,
	getValidationKeywordTerms,
} from "../../../i18n/validation-keywords.ts";
import { logger } from "../../../logger.ts";
import { shouldUnmuteRoomTemplate } from "../../../prompts.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
} from "../../../types/index.ts";
import { ModelType } from "../../../types/index.ts";
import {
	composePromptFromState,
	parseBooleanFromText,
	parseKeyValueXml,
} from "../../../utils.ts";

// Get text content from centralized specs
const spec = requireActionSpec("UNMUTE_ROOM");
const UNMUTE_TERMS = getValidationKeywordTerms("action.unmuteRoom.request", {
	includeAllLocales: true,
});

export const unmuteRoomAction: Action = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,
	examples: (spec.examples ?? []) as ActionExample[][],
	validate: async (runtime: IAgentRuntime, message: Memory) => {
		const text =
			typeof message?.content === "string"
				? message.content
				: (message?.content?.text ?? "");
		if (findKeywordTermMatch(text, UNMUTE_TERMS) === undefined) return false;
		const roomId = message.roomId;
		const roomState = await runtime.getParticipantUserState(
			roomId,
			runtime.agentId,
		);
		return roomState === "MUTED";
	},
	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		_callback?: HandlerCallback,
		_responses?: Memory[],
	): Promise<ActionResult> => {
		async function _shouldUnmute(state: State): Promise<boolean> {
			const shouldUnmutePrompt = composePromptFromState({
				state,
				template: shouldUnmuteRoomTemplate,
			});

			const response = await runtime.useModel(ModelType.TEXT_SMALL, {
				prompt: shouldUnmutePrompt,
				stopSequences: [],
			});

			const parsed = parseKeyValueXml<{ decision?: boolean | string }>(
				response,
			);
			const decisionValue = parsed?.decision ?? response.trim();
			const cleanedResponse = String(decisionValue).trim().toLowerCase();

			// Handle various affirmative responses
			if (
				parseBooleanFromText(decisionValue) ||
				cleanedResponse.includes("true") ||
				cleanedResponse.includes("yes")
			) {
				await runtime.createMemory(
					{
						entityId: message.entityId,
						agentId: message.agentId,
						roomId: message.roomId,
						content: {
							source: message.content.source,
							thought:
								"I will now unmute this room and start considering it for responses again",
							actions: ["UNMUTE_ROOM_STARTED"],
						},
					},
					"messages",
				);
				return true;
			}

			// Handle various negative responses
			if (
				cleanedResponse === "false" ||
				cleanedResponse === "no" ||
				cleanedResponse === "n" ||
				cleanedResponse.includes("false") ||
				cleanedResponse.includes("no")
			) {
				await runtime.createMemory(
					{
						entityId: message.entityId,
						agentId: message.agentId,
						roomId: message.roomId,
						content: {
							source: message.content.source,
							thought: "I tried to unmute a room but I decided not to",
							actions: ["UNMUTE_ROOM_FAILED"],
						},
					},
					"messages",
				);
				return false;
			}

			// Default to false if response is unclear
			logger.warn(
				{
					src: "plugin:advanced-capabilities:action:unmute_room",
					agentId: runtime.agentId,
					response,
				},
				"Unclear boolean response, defaulting to false",
			);
			return false;
		}

		if (!state) {
			return {
				text: "State is required for unmute room action",
				values: {
					success: false,
					error: "STATE_REQUIRED",
				},
				data: {
					actionName: "UNMUTE_ROOM",
					error: "State is required",
				},
				success: false,
				error: new Error("State is required for unmute room action"),
			};
		}

		const shouldUnmute = await _shouldUnmute(state);

		if (shouldUnmute) {
			try {
				await runtime.updateParticipantUserState(
					message.roomId,
					runtime.agentId,
					null,
				);

				const room = await runtime.getRoom(message.roomId);

				if (!room) {
					logger.warn(
						{
							src: "plugin:advanced-capabilities:action:unmute_room",
							agentId: runtime.agentId,
							roomId: message.roomId,
						},
						"Room not found",
					);
					return {
						text: `Room not found: ${message.roomId}`,
						values: {
							success: false,
							error: "ROOM_NOT_FOUND",
							roomId: message.roomId,
						},
						data: {
							actionName: "UNMUTE_ROOM",
							error: "Room not found",
							roomId: message.roomId,
						},
						success: false,
					};
				}

				// Ensure room has a name for consistent return values
				const roomName = room.name ?? `Room-${message.roomId.substring(0, 8)}`;

				await runtime.createMemory(
					{
						entityId: message.entityId,
						agentId: message.agentId,
						roomId: message.roomId,
						content: {
							thought: `I unmuted the room ${roomName}`,
							actions: ["UNMUTE_ROOM_START"],
						},
					},
					"messages",
				);

				return {
					text: `Room unmuted: ${roomName}`,
					values: {
						success: true,
						roomUnmuted: true,
						roomId: message.roomId,
						roomName: roomName,
						newState: "NONE",
					},
					data: {
						actionName: "UNMUTE_ROOM",
						roomId: message.roomId,
						roomName: roomName,
						unmuted: true,
					},
					success: true,
				};
			} catch (error) {
				logger.error(
					{
						src: "plugin:advanced-capabilities:action:unmute_room",
						agentId: runtime.agentId,
						error: error instanceof Error ? error.message : String(error),
					},
					"Error unmuting room",
				);
				return {
					text: "Failed to unmute room",
					values: {
						success: false,
						error: "UNMUTE_FAILED",
					},
					data: {
						actionName: "UNMUTE_ROOM",
						error: error instanceof Error ? error.message : String(error),
						roomId: message.roomId,
					},
					success: false,
					error: error instanceof Error ? error : new Error(String(error)),
				};
			}
		} else {
			return {
				text: "Decided not to unmute room",
				values: {
					success: true,
					roomUnmuted: false,
					roomId: message.roomId,
					reason: "CRITERIA_NOT_MET",
				},
				data: {
					actionName: "UNMUTE_ROOM",
					roomId: message.roomId,
					unmuted: false,
					reason: "Decision criteria not met",
				},
				success: true,
			};
		}
	},
};

File: packages/python/elizaos/features/advanced_capabilities/actions/unmute_room.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.types import Action, ActionExample, ActionResult, Content

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_action_spec("UNMUTE_ROOM")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    spec_examples = _spec.get("examples", [])
    if not isinstance(spec_examples, list):
        return []
    result: list[list[ActionExample]] = []
    for example in spec_examples:
        if not isinstance(example, list):
            continue
        row: list[ActionExample] = []
        for msg in example:
            if not isinstance(msg, dict):
                continue
            content = msg.get("content", {})
            text = ""
            actions: list[str] | None = None
            if isinstance(content, dict):
                text_val = content.get("text", "")
                text = str(text_val) if text_val else ""
                actions_val = content.get("actions")
                if isinstance(actions_val, list) and all(isinstance(a, str) for a in actions_val):
                    actions = list(actions_val)
            row.append(
                ActionExample(
                    name=str(msg.get("name", "")),
                    content=Content(text=text, actions=actions),
                )
            )
        if row:
            result.append(row)
    return result


@dataclass
class UnmuteRoomAction:
    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        room_id = message.room_id
        if not room_id:
            return False

        room = await runtime.get_room(room_id)
        if room is None:
            return False

        world_id = room.world_id
        if world_id:
            world = await runtime.get_world(world_id)
            if world and world.metadata:
                muted_rooms = world.metadata.get("mutedRooms", [])
                if str(room_id) in muted_rooms:
                    return True

        return False

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        room_id = message.room_id
        if not room_id:
            return ActionResult(
                text="No room specified to unmute",
                values={"success": False, "error": "no_room_id"},
                data={"actionName": "UNMUTE_ROOM"},
                success=False,
            )

        room = await runtime.get_room(room_id)
        if room is None:
            return ActionResult(
                text="Room not found",
                values={"success": False, "error": "room_not_found"},
                data={"actionName": "UNMUTE_ROOM"},
                success=False,
            )

        room_name = str(room.name) if room.name else "Unknown Room"

        world_id = room.world_id
        if world_id:
            world = await runtime.get_world(world_id)
            if world and world.metadata:
                muted_rooms = list(world.metadata.get("mutedRooms", []))
                room_id_str = str(room_id)

                if room_id_str in muted_rooms:
                    muted_rooms.remove(room_id_str)
                    world.metadata["mutedRooms"] = muted_rooms
                    await runtime.update_world(world)

        await runtime.create_memory(
            content=Content(
                text=f"Unmuted room: {room_name}",
                actions=["UNMUTE_ROOM"],
            ),
            room_id=room_id,
            entity_id=runtime.agent_id,
            memory_type="action",
            metadata={"type": "UNMUTE_ROOM", "roomName": room_name},
        )

        response_content = Content(
            text=f"I have unmuted {room_name}. I will now respond to messages there.",
            actions=["UNMUTE_ROOM"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Unmuted room: {room_name}",
            values={
                "success": True,
                "unmuted": True,
                "roomId": str(room_id),
                "roomName": room_name,
            },
            data={
                "actionName": "UNMUTE_ROOM",
                "roomId": str(room_id),
                "roomName": room_name,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


unmute_room_action = Action(
    name=UnmuteRoomAction.name,
    similes=UnmuteRoomAction().similes,
    description=UnmuteRoomAction.description,
    validate=UnmuteRoomAction().validate,
    handler=UnmuteRoomAction().handler,
    examples=UnmuteRoomAction().examples,
)
